src/lib/core/fiber.c

```diff
@@ -502,14 +502,13 @@ void
 fiber_wakeup(struct fiber *f)
 {
 	/*
-	 * DEAD is checked both in the assertion and in the release build
-	 * because it should not ever happen, at least internally. But in some
-	 * user modules it might happen, and better ignore such fibers.
-	 * Especially since this was allowed for quite some time in the public
-	 * API and need to keep it if it costs nothing, for backward
-	 * compatibility.
+	 * DEAD fiber can be lingering in the cord fiber list
+	 * if it is joinable. And once its execution is complete
+	 * it should be reaped with fiber_join() call.
+	 *
+	 * Still our API allows to call fiber_wakeup() on dead
+	 * fibers so simply ignore it.
```

Comment:
Please, keep the assertion but add a check into it that the fiber can be joinable. The assertion was here for a reason - to catch bugs in our code. Removal of the assertion reduces probability of catching a bug. It is still dangerous to wakeup a dead non-joinable fiber because it can be reused any moment and you will wakeup an irrelevant fiber. I really don't understand your hate towards assertions. The new assertion should be `assert((f->flags & FIBER_IS_DEAD) == 0 || (f->flags & FIBER_IS_JOINABLE) != 0);`.

Reply:
updated